[Keyboard] Pink Labs e88

2-n opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description
Adding the keymap for the Pink Labs e88 PCB, for the Bliss keyboard.

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

ridingqwerty commented on keyboards/e88-f13/rules.mk:
```diff
@@ -0,0 +1,33 @@
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
+FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
+HD44780_ENABLE = no         # Enable support for HD44780 based LCDs
```
```suggestion
```

ridingqwerty commented on keyboards/e88-f13/readme.md:
```diff
@@ -0,0 +1,21 @@
+# e88-f13
```
```suggestion
# e88_f13
```

ridingqwerty commented on keyboards/e88-f13/readme.md:
````diff
@@ -0,0 +1,21 @@
+Make example for this keyboard (after setting up your build environment):
+
+```sh
+make e88-f13:default
````
```suggestion
make e88_f13:default
```

ridingqwerty commented on keyboards/e88-f13/readme.md:
````diff
@@ -0,0 +1,21 @@
+Example of flashing this keyboard:
+
+```sh
+make e88-f13:default:flash
````
```suggestion
make e88_f13:default:flash
```

ridingqwerty commented on keyboards/e88-f13/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,42 @@
+// entirely and just use numbers.
+
+#define _MA 0 //Main layer
+#define _FN 1 //Main layer
```
```suggestion
#define _FN 1 //Function layer

```

ridingqwerty commented on keyboards/e88-f13/config.h:
```diff
@@ -0,0 +1,251 @@
+#define DEVICE_VER      0x0001
+#define MANUFACTURER    Pink Labs
+#define PRODUCT         F13
+#define DESCRIPTION     A TKL
```
```suggestion
```

ridingqwerty commented on keyboards/e88-f13/config.h:
```diff
@@ -0,0 +1,251 @@
+//#define MIDI_BASIC
+
+/* enable advanced MIDI features:
+   - MIDI notes can be added to the keymap
+   - Octave shift and transpose
+   - Virtual sustain, portamento, and modulation wheel
+   - etc.
+*/
+//#define MIDI_ADVANCED
+
+/* override number of MIDI tone keycodes (each octave adds 12 keycodes and allocates 12 bytes) */
+//#define MIDI_TONE_KEYCODE_OCTAVES 1
+
+/*
+ * HD44780 LCD Display Configuration
+ */
+/*
+#define LCD_LINES           2     //< number of visible lines of the display
+#define LCD_DISP_LENGTH    16     //< visibles characters per line of the display
+
+#define LCD_IO_MODE      1            //< 0: memory mapped mode, 1: IO port mode
+
+#if LCD_IO_MODE
+#define LCD_PORT         PORTB        //< port for the LCD lines
+#define LCD_DATA0_PORT   LCD_PORT     //< port for 4bit data bit 0
+#define LCD_DATA1_PORT   LCD_PORT     //< port for 4bit data bit 1
+#define LCD_DATA2_PORT   LCD_PORT     //< port for 4bit data bit 2
+#define LCD_DATA3_PORT   LCD_PORT     //< port for 4bit data bit 3
+#define LCD_DATA0_PIN    4            //< pin for 4bit data bit 0
+#define LCD_DATA1_PIN    5            //< pin for 4bit data bit 1
+#define LCD_DATA2_PIN    6            //< pin for 4bit data bit 2
+#define LCD_DATA3_PIN    7            //< pin for 4bit data bit 3
+#define LCD_RS_PORT      LCD_PORT     //< port for RS line
+#define LCD_RS_PIN       3            //< pin  for RS line
+#define LCD_RW_PORT      LCD_PORT     //< port for RW line
+#define LCD_RW_PIN       2            //< pin  for RW line
+#define LCD_E_PORT       LCD_PORT     //< port for Enable line
+#define LCD_E_PIN        1            //< pin  for Enable line
+#endif
+*/
```
```suggestion
```

2-n commented:
Fixed everything that was of issue and removed unnecessary things from config like you stated. Didn't realize files needed to be same name as folder was confused myself, but everything was renamed to prevent any further issue and now no problems with either check.

ridingqwerty commented on keyboards/e88/config.h:
```diff
@@ -0,0 +1,162 @@
+#define PRODUCT_ID      0x0187
+#define DEVICE_VER      0x0001
+#define MANUFACTURER    Pink Labs
+#define PRODUCT         Bliss
```
Why "Bliss" here and not "e88" or the like?

2-n replied:
PCB was designed for my board Bliss.

zvecr commented on keyboards/e88/readme.md:
````diff
@@ -0,0 +1,21 @@
+* Hardware Supported: e88 f13 PCB (proto and final rev)
+* Hardware Availability: [labs.pink](http://www.labs.pink/)
+
+Make example for this keyboard (after setting up your build environment):
+
+```sh
+make e88:default
+```
+
+Example of flashing this keyboard:
+
+```sh
+make e88:default:flash
+```
+
+See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools) and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more information. Brand new to QMK? Start with our [Complete Newbs Guide](https://docs.qmk.fm/#/newbs).
````
```suggestion
Make example for this keyboard (after setting up your build environment):

    make e88:default

Flashing example for this keyboard:

    make e88:default:flash

See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools) and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more information. Brand new to QMK? Start with our [Complete Newbs Guide](https://docs.qmk.fm/#/newbs).
```

zvecr commented on keyboards/e88/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,42 @@
+ */
+#include QMK_KEYBOARD_H
+
+// Each layer gets a name for readability, which is then used in the keymap matrix below.
+// The underscores don't mean anything - you can have a layer called STUFF or any other name.
+// Layer names don't all need to be of the same length, obviously, and you can also skip them
+// entirely and just use numbers.
+
+#define _MA 0 //Main layer
+#define _FN 1 //Function layer
```
```suggestion
// Defines names for use in layer keycodes and the keymap
enum layer_names {
    _MA, //Main layer
    _FN  //Function layer
};
```

zvecr commented on keyboards/e88/rules.mk:
```diff
@@ -0,0 +1,31 @@
+# MCU name
+MCU = atmega32u4
+
+# Bootloader selection
+#   Teensy       halfkay
+#   Pro Micro    caterina
+#   Atmel DFU    atmel-dfu
+#   LUFA DFU     lufa-dfu
+#   QMK DFU      qmk-dfu
+#   ATmega32A    bootloadHID
+#   ATmega328P   USBasp
+BOOTLOADER = qmk-dfu
+
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
+MOUSEKEY_ENABLE = no        # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = yes        # Console for debug
+COMMAND_ENABLE = yes        # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = NO        # Enable keyboard RGB underglow
+MIDI_ENABLE = no            # MIDI support
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
```
```suggestion
# Bootloader selection
BOOTLOADER = qmk-dfu

# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = lite     # Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = no            # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
```

zvecr commented on keyboards/e88/config.h:
```diff
@@ -0,0 +1,162 @@
+//#define MAGIC_KEY_LAYER0         0
+//#define MAGIC_KEY_LAYER0_ALT     GRAVE
+//#define MAGIC_KEY_LAYER1         1
+//#define MAGIC_KEY_LAYER2         2
+//#define MAGIC_KEY_LAYER3         3
+//#define MAGIC_KEY_LAYER4         4
+//#define MAGIC_KEY_LAYER5         5
+//#define MAGIC_KEY_LAYER6         6
+//#define MAGIC_KEY_LAYER7         7
+//#define MAGIC_KEY_LAYER8         8
+//#define MAGIC_KEY_LAYER9         9
+//#define MAGIC_KEY_BOOTLOADER     B
+//#define MAGIC_KEY_BOOTLOADER_ALT ESC
+//#define MAGIC_KEY_LOCK           CAPS
+//#define MAGIC_KEY_EEPROM         E
+//#define MAGIC_KEY_EEPROM_CLEAR   BSPACE
+//#define MAGIC_KEY_NKRO           N
+//#define MAGIC_KEY_SLEEP_LED      Z
+
+/*
+ * Feature disable options
+ *  These options are also useful to firmware size reduction.
+ */
+
+/* disable debug print */
+//#define NO_DEBUG
+
+/* disable print */
+//#define NO_PRINT
+
+/* disable action features */
+//#define NO_ACTION_LAYER
+//#define NO_ACTION_TAPPING
+//#define NO_ACTION_ONESHOT
+//#define NO_ACTION_MACRO
+//#define NO_ACTION_FUNCTION
+
+/* Bootmagic Lite key configuration */
+// #define BOOTMAGIC_LITE_ROW 0
+// #define BOOTMAGIC_LITE_COLUMN 0
```
```suggestion
/* If defined, GRAVE_ESC will always act as ESC when CTRL is held.
 * This is useful for the Windows task manager shortcut (ctrl+shift+esc).
 */
//#define GRAVE_ESC_CTRL_OVERRIDE

/*
 * Force NKRO
 *
 * Force NKRO (nKey Rollover) to be enabled by default, regardless of the saved
 * state in the bootmagic EEPROM settings. (Note that NKRO must be enabled in the
 * makefile for this to work.)
 *
 * If forced on, NKRO can be disabled via magic key (default = LShift+RShift+N)
 * until the next keyboard reset.
 *
 * NKRO may prevent your keystrokes from being detected in the BIOS, but it is
 * fully operational during normal computer usage.
 *
 * For a less heavy-handed approach, enable NKRO via magic key (LShift+RShift+N)
 * or via bootmagic (hold SPACE+N while plugging in the keyboard). Once set by
 * bootmagic, NKRO mode will always be enabled until it is toggled again during a
 * power-up.
 *
 */
//#define FORCE_NKRO

/*
 * Feature disable options
 *  These options are also useful to firmware size reduction.
 */

/* disable debug print */
//#define NO_DEBUG

/* disable print */
//#define NO_PRINT

/* disable action features */
//#define NO_ACTION_LAYER
//#define NO_ACTION_TAPPING
//#define NO_ACTION_ONESHOT

/* disable these deprecated features by default */
#define NO_ACTION_MACRO
#define NO_ACTION_FUNCTION
```

noroadsleft commented on keyboards/e88/rules.mk:
```diff
@@ -0,0 +1,31 @@
+RGBLIGHT_ENABLE = NO        # Enable keyboard RGB underglow
+MIDI_ENABLE = no            # MIDI support
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
```
```suggestion
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
```

noroadsleft commented on keyboards/e88/rules.mk:
```diff
@@ -0,0 +1,31 @@
+MCU = atmega32u4
+
+# Bootloader selection
+#   Teensy       halfkay
+#   Pro Micro    caterina
+#   Atmel DFU    atmel-dfu
+#   LUFA DFU     lufa-dfu
+#   QMK DFU      qmk-dfu
+#   ATmega32A    bootloadHID
+#   ATmega328P   USBasp
```
```suggestion
```

2-n commented:
> 
> 
> I'm finding myself uncomfortable with the fact that the keyboard name doesn't match the USB Product string, but I'm at a loss for a suggestion here.

Ended up changing the name to e88 in the USB Product string and adding a description. 

> 
> 
> Whoops, still missing an info.json. Otherwise LGTM

I reviewed others and I did as best of a job as I could to mimic for my board, but I think I did everything correctly. If there is anything wrong I am glad to fix it.

2-n commented:
> 
> 
> LAYOUT_all is fine, but support for the tkl_ansi and tkl_iso community layouts should really be added.
> Also a info.json is missing, documented here https://docs.qmk.fm/#/reference_info_json

Not sure how I would go about doing everything for the community layouts as this is DK Saver layout PCB with an F13 key. Though in the info.json has support for configuring all possible layouts in the one layout I have.

noroadsleft commented on keyboards/e88/config.h:
```diff
@@ -23,7 +23,8 @@
 #define MANUFACTURER    Pink Labs
-#define PRODUCT         Bliss
+#define PRODUCT         e88
+#define DESCRIPTION     e88 PCB for Bliss F13 TKL
```
The `DESCRIPTION` string is actually deprecated, and no longer used anywhere within QMK.

```suggestion
```

noroadsleft commented on keyboards/e88/info.json:
```diff
@@ -0,0 +1,109 @@
+    "width": 17,
+    "height": 6,
+    "layouts": {
+        "LAYOUT": {
```
File contents are fine here, but this string needs to be an exact match for the name of the layout macro to which it applies:

https://github.com/qmk/qmk_firmware/blob/a88884870160f620c1e9161d32982701de301864/keyboards/e88/e88.h#L30

```suggestion
        "LAYOUT_all": {
```

noroadsleft commented on keyboards/e88/readme.md:
```diff
@@ -0,0 +1,17 @@
+
+![pink labs logo](https://cdn.shopify.com/s/files/1/0302/4198/0548/files/logo_x60.png?v=1579122787)
+
+* Maintainer: [elli](https://github.com/2-n)
```
We can call out the e88/Bliss connection here.

```suggestion
A tenkeyless PCB featuring the DK Saver layout, as designed for the Bliss keyboard by Pink Labs.

* Maintainer: [elli](https://github.com/2-n)
```

noroadsleft commented:
> Not sure how I would go about doing everything for the community layouts as this is DK Saver layout PCB with an F13 key. Though in the info.json has support for configuring all possible layouts in the one layout I have.

Yeah, you're fine here. None of the community layouts support an F13 key.

noroadsleft commented on keyboards/e88/info.json:
```diff
@@ -0,0 +1,109 @@
+    "keyboard_name": "e88",
+    "url": "",
+    "maintainer": "2-n",
+    "width": 17,
+    "height": 6,
```
```suggestion
    "width": 18.25,
    "height": 6.25,
```

2-n commented:
> Everything besides this looks good to me.

Sweet, thanks. Wasn't sure if it should be the same as the matrix or the physical units.

noroadsleft commented:
> Sweet, thanks. Wasn't sure if it should be the same as the matrix or the physical units.

Yeah, it's supposed to be physical units.

noroadsleft commented:
Thanks!

For future reference, we recommend against committing to your `master` branch as you've done here, because pull requests from modified `master` branches can make it more difficult to keep your QMK fork updated. It is highly recommended for QMK development – regardless of what is being done or where – to keep your master updated, but **NEVER** commit to it. Instead, do all your changes in a branch (branches are basically free in Git) and issue PRs from your branches when you're developing.

There are instructions on how to keep your fork updated here:

[**Best Practices: Your Fork's Master: Update Often, Commit Never**](https://docs.qmk.fm/#/newbs_git_using_your_master_branch)

[Fixing Your Branch](https://docs.qmk.fm/#/newbs_git_resynchronize_a_branch) will walk you through fixing up your `master` branch moving forward. If you need any help with this just ask.
